cobra/core/Reaction.py

```diff
@@ -154,26 +158,24 @@ def _reset_var_cache(self):
 
     @property
     def objective_coefficient(self):
-        """ Get or set the objective coefficient (float)
+        """ Get the coefficient for this reaction in a linear
+        objective (float)
+
+        Assuming that the objective of the associated model is summation of
+        fluxes from a set of reactions, the coefficient for each reaction
+        can be obtained individually using this property. A more general way
+        is to use the `model.objective` property directly.
         """
-        if self.model is not None and self.model.solver.objective is not None:
-            coefficients_dict = \
-                self.model.solver.objective.expression.as_coefficients_dict()
-            forw_coef = coefficients_dict.get(self.forward_variable, 0)
-            rev_coef = coefficients_dict.get(self.reverse_variable, 0)
-            if forw_coef == -rev_coef:
-                self._objective_coefficient = float(forw_coef)
-            else:
-                self._objective_coefficient = 0
-        return self._objective_coefficient
+        return linear_reaction_coefficients(self.model, [self]).get(self, 0)
 
     @objective_coefficient.setter
     def objective_coefficient(self, value):
-        if self.model is not None:
-            coef_difference = value - self.objective_coefficient
-            self.model.solver.objective += \
-                coef_difference * self.flux_expression
-        self._objective_coefficient = value
+        if self.model is None:
+            raise AttributeError('cannot assign objective to a missing model')
+        if self.flux_expression is not None:
+            self.model.solver.objective.set_linear_coefficients(
```

Comment:
That should use `set_objective_*` from the solver utils so we don't have to worry about the context manager.